src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -96,28 +96,31 @@ final class HashMap[K, +V] private[immutable] (private[immutable] val rootNode:
     rootNode.getOrElse(key, keyUnimprovedHash, keyHash, 0, default)
   }
 
+  @`inline` private[this] def newHashMapOrThis[V1 >: V](newRootNode: MapNode[K, V1]): HashMap[K, V1] =
+    if (newRootNode eq rootNode) this else new HashMap(newRootNode)
+
   def updated[V1 >: V](key: K, value: V1): HashMap[K, V1] = {
     val keyUnimprovedHash = key.##
-    val keyHash = improve(keyUnimprovedHash)
-    val newRootNode = rootNode.updated(key, value, keyUnimprovedHash, keyHash, 0)
-    if (newRootNode ne rootNode) new HashMap(newRootNode) else this
+    newHashMapOrThis(rootNode.updated(key, value, keyUnimprovedHash, improve(keyUnimprovedHash), 0))
   }
 
   def removed(key: K): HashMap[K, V] = {
     val keyUnimprovedHash = key.##
-    val keyHash = improve(keyUnimprovedHash)
-    val newRootNode = rootNode.removed(key, keyUnimprovedHash, keyHash, 0)
-    if (newRootNode ne rootNode) new HashMap(newRootNode) else this
+    newHashMapOrThis(rootNode.removed(key, keyUnimprovedHash, improve(keyUnimprovedHash), 0))
   }
 
   override def concat[V1 >: V](that: scala.IterableOnce[(K, V1)]): HashMap[K, V1] = that match {
-    case hm: HashMap[K, V1] =>
-      if (hm.rootNode eq rootNode) this else {
-        val newRootNode = rootNode.concat(hm.rootNode, 0)
-        if (newRootNode eq rootNode) this
-        else if (newRootNode eq hm.rootNode) hm
-        else new HashMap(newRootNode)
+    case hm: HashMap[K, V1] => newHashMapOrThis(rootNode.concat(hm.rootNode, 0))
+    case hm: collection.mutable.HashMap[K, V] =>
+      val iter = hm.nodeIterator
+      var newRootNode = rootNode
+      while (iter.hasNext) {
+        val next = iter.next()
+        val originalHash = hm.unimproveHash(next.hash)
+        val hash = improve(originalHash)
```

Comment:
I wonder how well the simpler improvement algorithm would for for CHAMP. If we could use it there, you wouldn't have to unimprove and improve the hash.

Reply:
Good question, not sure how to evaluate which one is better.

Reply:
It's a lousy algorithm if you don't have good hash values to begin with.  For instance, if I have `0 until 16384 by 256`, it's basically all collisions in the bits that matter.